File: nvim/language-examples/python/shapes.py
import math
from abc import ABC, abstractmethod

# go to definition on `Shape`, `Circle`, `Rectangle` in main.py jumps here
# hover on any class or method shows its docstring and signature

class Shape(ABC):
    """Base class for all shapes.

    All shapes must implement the area() method.
    This docstring is shown when hovering over Shape anywhere.
    """

    @abstractmethod
    def area(self) -> float:
        """Calculate and return the area of the shape.

        Returns:
            float: The area of the shape.

        Hover on any call to area() to see this docstring.
        Go to definition on `area` call in main.py jumps here.
        """
        ...


class Circle(Shape):
    """A circle defined by its radius.

    Hover over Circle in main.py to see this docstring.
    """

    def __init__(self, radius: float) -> None:
        """Initialize a Circle with the given radius.

        Args:
            radius: The radius of the circle. Must be positive.

        Hover on `Circle(5.0)` in main.py shows this signature and docstring.
        """
        self.radius = radius

    def area(self) -> float:
        """Return the area of the circle.

        Returns:
            float: pi * radius^2

        Find references on `area` shows all call sites across files.
        """
        return math.pi * self.radius ** 2


class Rectangle(Shape):
    """A rectangle defined by width and height."""

    def __init__(self, width: float, height: float) -> None:
        """Initialize a Rectangle.

        Args:
            width: The horizontal dimension.
            height: The vertical dimension.
        """
        self.width = width
        self.height = height

    def area(self) -> float:
        """Return the area of the rectangle.

        Returns:
            float: width * height
        """
        return self.width * self.height


def add(a: int, b: int) -> int:
    """Return the sum of two integers.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        int: The sum of a and b.

    Hover on any call to add() in main.py to see this docstring and signature.
    """
    return a + b

File: nvim/language-examples/fortran/shapes.f90
module shapes
    implicit none

    ! go to definition on `shape_t` or `circle_t` in main.f90 jumps here
    ! hover on any type shows its components

    type, abstract :: shape_t
    contains
        ! go to definition on `area` call in main.f90 jumps here
        procedure(area_interface), deferred :: area
    end type

    abstract interface
        function area_interface(self) result(a)
            import shape_t
            class(shape_t), intent(in) :: self
            real :: a
        end function
    end interface

    type, extends(shape_t) :: circle_t
        real :: radius
    contains
        ! find references on `area` shows all implementations and call sites
        procedure :: area => circle_area
    end type

    type, extends(shape_t) :: rectangle_t
        real :: width, height
    contains
        procedure :: area => rectangle_area
    end type

contains

    function circle_area(self) result(a)
        class(circle_t), intent(in) :: self
        real :: a
        a = 3.14159265 * self%radius ** 2
    end function

    function rectangle_area(self) result(a)
        class(rectangle_t), intent(in) :: self
        real :: a
        a = self%width * self%height
    end function

    integer function add(a, b)
        integer, intent(in) :: a, b
        add = a + b
    end function

end module shapes
